python/ray/external_storage.py

```diff
@@ -1,8 +1,98 @@
 import abc
 import os
+import re
+from collections import namedtuple
 from typing import List
 
 import ray
+from ray.ray_constants import DEFAULT_OBJECT_PREFIX
+
+
+class ExternalStorageURLProtocol:
+    """Class to parse/generate urls for object spilling.
+
+    There will be 2 types of URLs.
+    1. URL used internally. These urls are stored in GCS.
+    2. URL used for external storage.
+
+    This class is used as a translator layer between 1 and 2.
+    """
+
+    def __init__(self, prefix: str):
+        self._prefix = prefix
+        # NOTE: To optimize the performance, Ray fusions small
+        # objects when it stored them. object_id_stored is
+        # an object_id that is used as a filename.
```

Comment:
Do we need object_id_stored? It can just be the first object id, or "combined-".

Reply:
I probably didn't understand your suggestion here. My thought here was that we need the object_id_stored (btw, this is the bad name. I should probably rename it to object_id_in_filename or something) when we restore the object. For example, if URLWithOffset doesn't contain the object_id_stored, we don't know how to get the `base_url` from the `URLWithOffset`.

Reply:
Won't the URL be enough to identify the file / offset?

Reply:
The filename doesn't need the object id at all really, the file names can be random